plugins/uefi-dbx/fu-uefi-dbx-device.c

```diff
@@ -33,6 +69,14 @@ fu_uefi_dbx_device_write_firmware(FuDevice *device,
 	if (fw == NULL)
 		return FALSE;
 
+#ifdef WITH_UEFI_DBX_SNAPD_NOTIFIER
+	g_debug("firmware write - notify snapd");
+	if (!fu_uefi_dbx_device_maybe_notify_snapd_prepare(FU_UEFI_DBX_DEVICE(device), fw, error)) {
+		g_prefix_error(error, "cannot notify snapd of pending update");
+		return FALSE;
+	}
+#endif
```

Comment:
shouldn't this be in the prepare firmware callback?  Or is write really the correct place?

Reply:
That was another option. The upside of `_write()` is that we're really committed to doing the update at this point. I can try to move that to prepare if it makes more sense. AFAICT the stream held by FuFirmware is seekable, so `fu_firmware_get_bytes()` should work then called in both.

Reply:
That's done in `device_cleanup()`, which AFAIU is called eventually when write fails.

However, starting the DBX update operation has some additional side effects on the snapd side, for instance snaps which are part of the boot process (or have artifacts which are measured) will be blocked form updating. Thus, the cleanup is what would be executed in the 'happy' path. Unexpected reboots or a crash in fwupd should result in `fu_uefi_dbx_snapd_notifier_dbx_manager_startup` being called at some point, which triggers an error handling path if an update was started and not finished. If all fails, there's incomplete work objects in snapd (aka. changes, dbx update is tracked as one) get aborted after a generous waiting time, which would unblock everything again. Obviously, once `prepare` was invoked, a subsequent `prepare` will result in an error, if cleanup/startup wasn't called in between.